Add documentation for error handling (#147)
Shopify/ui-extensions: documentation/error-handling.md

# Error handling

> Because ui extensions script runs in non-browser environment (ex: WebWorker in Web, JavascriptCore on Mobile), it is recommended to listen and send errors to server manually. However, it is still possible to use external libraries for reporting errors. This document shows a couple of examples about how you can handle it.

Reviewer: ```suggestion
Because UI extension scripts run in non-browser environments (Web Worker in web, JavascriptCore on Mobile), it is recommended to manually set up error listeners and error reporting. However, it is possible to use external libraries for error reporting.
```

> ## The generic way of handling errors

Reviewer: ```suggestion
## Error handling in UI extensions
```

> There are two types of errors when using ui extensions: lifecycle and non-lifecycle error.

Reviewer: ```suggestion
UI extensions can throw two types of error: lifecycle errors and non-lifecycle errors.
```

> - **Lifecycle error:** most errors will fall into this type. It could happen when ui extensions script is loaded / imported, the script is run / rendered, or any interaction via callback of ui extensions components or apis. To catch this error, you can add event listener for `unhandledrejection`.

Reviewer: ```suggestion
### Lifecycle error

Most errors will fall into this type. These errors can be thrown when:

- a UI extension script is loaded or imported
- the UI extension script is run
- during any callback interaction with UI extension components or APIs

To catch these errors, add an event listener for `unhandledrejection`.
```

  ```javascript
  self.onunhandledrejection = (error) => {
    console.warn('onunhandledrejection', error);
  };
  // or
  self.addEventListener('unhandledrejection', (error) => {
    console.warn('event unhandledrejection', error);
  });
  ```

> - **Non-lifecycle error:** this error could happen when the script throw exception while running setTimeout callback. To catch this error, you can add event listener for `error`.

Reviewer: ```suggestion
### Non-lifecycle error

These errors can be thrown when the script encounters an exception while running a `setTimeout` callback. To catch these errors, add an event listener for `error`.
```

Author: This is what I observe. Not sure if there are more cases :D

  ```javascript
  self.onerror = (error) => {
    console.warn('onerror', error);
  };
  // or
  self.addEventListener('error', (error) => {
    console.warn('event error', error);
  });
  ```

## Handle errors via 3rd-party libraries

### Bugsnag

See https://docs.bugsnag.com/platforms/javascript/.

> You can use either CDN version or npm dependencies to run Bugsnag. However, at runtime, Bugsnag requires additional polyfill as it tries to access browser environment. Below script has been tested at the time of writing this docs

Reviewer: ```suggestion
You can run Bugsnag using the CDN version or npm dependencies. However, Bugsnag requires additional polyfills at runtime, as it will try to access the browser environment. The script below has been tested at the time of writing these docs.
```

```javascript
/**
 * Window polyfill
 */
(() => {
  const document = {
    documentElement: {
      outerHTML: '',
      createElement: () => ({}),
      clientWidth: 0,
      clientHeight: 0,
    },
    addEventListener: () => {},
  };
  const history = {};
  const window = self;
  self.window = window;
  window.document = document;
  window.history = history;
})();

/**
 * Bugsnag
 */
// Add bugsnag CDN javascript here
Bugsnag.start({apiKey: 'YOUR-BUGSNAG-API-KEY'});

/**
 * Add error handlers
 */
self.onerror = (error) => {
  Bugsnag.notify(error);
};
```

> By default, Bugsnag has already sent error for `unhandledrejection`. You only need to listen to `onerror`.

Reviewer: ```suggestion
By default, Bugsnag will listen for `unhandledrejection`. You only need to manually listen to `onerror`.
```

### Sentry

See https://docs.sentry.io/platforms/javascript/.

> Using Sentry for ui extensions is very straightforward. You only need Sentry main bundle or `@sentry/browser`.

Reviewer: ```suggestion
Using Sentry for UI extensions is straightforward. You only need to import the main Sentry bundle or `@sentry/browser` No polyfills are required.
```

```javascript
// Add Sentry CDN javascript here
self.Sentry = Sentry;
Sentry.init({
  dsn: 'YOUR-SENTRY-DSN-VALUE',
});

/**
 * Add error handlers
 */
self.onerror = (error) => {
  Sentry.captureException(new Error(error.reason.stack));
};

self.onunhandledrejection = (error) => {
  Sentry.captureException(new Error(error.reason.stack));
};
```

> By default, Sentry does not send any error. You need to add listeners for both `error` and `unhandledRejection`.

Reviewer: ```suggestion
By default, Sentry does not listen for any errors. You will need to manually add listeners for both `error` and `unhandledRejection`.
```

> # Error handling

Reviewer 2: We should maybe add a quick note/ links for connecting react error boundaries to these services for developers writing react extensions

Author: Are you thinking about using ErrorBoundary for React? I don't think it can catch everything, like importScripts... I don't think there is anything specific to React that we could highlight here.

Reviewer 2: I was just thinking of a one-sentence thing, something like "If you are writing your UI extension in React, you can follow Bugsnag’s [React integration guide](https://docs.bugsnag.com/platforms/javascript/legacy/react/) to get additional context on errors thrown while rendering your components."

> Because UI extension scripts run in non-browser environments (Web Worker in web, JavascriptCore on Mobile), it is recommended to manually set up error listeners and error reporting. However, it is possible to use external libraries for error reporting.

Reviewer 2: I'm not sure what this first sentence is telling me I should be doing.

## Error handling in UI extensions

UI extensions can throw two types of error: lifecycle errors and non-lifecycle errors.

### Lifecycle error

Most errors will fall into this type. These errors can be thrown when:

- a UI extension script is loaded or imported
- the UI extension script is run
- during any callback interaction with UI extension components or APIs

To catch these errors, add an event listener for `unhandledrejection`.

  ```javascript
  self.onunhandledrejection = (error) => {
    console.warn('onunhandledrejection', error);
  };
  // or
  self.addEventListener('unhandledrejection', (error) => {
    console.warn('event unhandledrejection', error);
  });
  ```

> ### Non-lifecycle error

Reviewer 2: I don't think we should break these down this way. They do technically happen at these times, but that is just a factor of our implementation. For a developer, what they need to know is just that they can add the same global event listeners for untaught exceptions (the error event) and for unhandled promise rejections, just like they can in other JS environments.

These errors can be thrown when the script encounters an exception while running a `setTimeout` callback. To catch these errors, add an event listener for `error`.

  ```javascript
  self.onerror = (error) => {
    console.warn('onerror', error);
  };
  // or

>   self.addEventListener('error', (error) => {

Reviewer 2: Can we pick one of the formats and recommend only that one? I think we should just do the more modern addEventListener() examples.

    console.warn('event error', error);
  });
  ```

> ## Handle errors via 3rd-party libraries

Reviewer 2: This is more about reporting than handling

Reviewer 2: Also, what's our recommendation for someone who wants to know they have errors? I think we should probably just recommend the one with the smallest bundle size, though if that's bugsnag, the required polyfill is pretty terrible...

### Bugsnag

> See https://docs.bugsnag.com/platforms/javascript/.

Reviewer 2: Do I have to go here? Are the instructions in this doc not enough? I'd rather we give them all the common answers they need here, and link them to the docs as an afterthought if they have special cases.

> You can run Bugsnag using the CDN version or npm dependencies. However, Bugsnag requires additional polyfills at runtime, as it will try to access the browser environment. The script below has been tested at the time of writing these docs.

Reviewer 2: How would they use the CDN version? If they can use both, which do we prefer? Can we show them the commands and code they need for the preferred option?

```javascript
/**

>  * Window polyfill

Reviewer 2: Is there an issue on bugsnag for a version that could work in a worker? If not, could we open one? It's very unfortunate that developers need to do this, since they are changing the global environment of a sandbox they are not supposed to control. Maybe we should just add this in our sandbox instead?

 */
(() => {
  const document = {
    documentElement: {
      outerHTML: '',
      createElement: () => ({}),
      clientWidth: 0,
      clientHeight: 0,
    },
    addEventListener: () => {},
  };
  const history = {};
  const window = self;
  self.window = window;
  window.document = document;
  window.history = history;
})();

/**
 * Bugsnag
 */
// Add bugsnag CDN javascript here
Bugsnag.start({apiKey: 'YOUR-BUGSNAG-API-KEY'});

/**
 * Add error handlers
 */
self.onerror = (error) => {
  Bugsnag.notify(error);
};
```

> By default, Bugsnag will listen for `unhandledrejection`. You only need to manually listen to `onerror`.

Reviewer 2: Why is this, it seems very unexpected to me. The unhandled rejections are newer, I would have thought for sure bugsnag listened for errors automatically.

### Sentry

See https://docs.sentry.io/platforms/javascript/.

> Using Sentry for UI extensions is straightforward. You only need to import the main Sentry bundle or `@sentry/browser` No polyfills are required.

Reviewer 2: Is "import the main Sentry bundle" talking about loading from a CDN again? I don't know what it is describing compared to the NPM package.

```javascript
// Add Sentry CDN javascript here

> self.Sentry = Sentry;

Reviewer 2: Does it need to be a global?

Sentry.init({
  dsn: 'YOUR-SENTRY-DSN-VALUE',
});

/**
 * Add error handlers
 */
self.onerror = (error) => {
  Sentry.captureException(new Error(error.reason.stack));
};

self.onunhandledrejection = (error) => {
  Sentry.captureException(new Error(error.reason.stack));
};
```

> By default, Sentry does not listen for any errors. You will need to manually add listeners for both `error` and `unhandledRejection`.

Reviewer 2: This is equally odd to me as for bugsnag. Do they also not listen by default in a browser context?

# Error handling

> UI extensions support standard web errors. You can add `unhandledrejection` listener for unhandled promise rejections or `error` listener for other exceptions like Javascript runtime error or a resource fails to load.

Reviewer 3: ```suggestion
UI extensions support standard web errors. You can add an `unhandledrejection` listener for unhandled promise rejections or an `error` listener for other exceptions like JavaScript runtime errors or resources that fail to load.
```

> UI extensions support standard web errors. You can add an `unhandledrejection` listener for unhandled promise rejections or an `error` listener for other exceptions like JavaScript runtime errors or resources that fail to load.

Reviewer 2: I forget, did we actually implement these for the native JS sandbox as well?

Reviewer 2: Can we add a quick recommendation here? Something like, "If you don’t already have an error reporting tool, but you want to be alerted when errors occur in your extension, we recommend X. It has a free version, and has the smallest bundle size of the error tracking services documented in this guide."

Author: > did we actually implement these for the native JS sandbox as well?

By `native JS sandbox`, you mean mobile, don't you? If so, no. Current Android gets this for free because it switch to use WebView like Web. I haven't verified it in current iOS as it is using JavascriptCore. 

For future Android and iOS powered by React Native in Mobile Shopify and POS, we get it for free as they will be both using WebView (hidden) and WebWorker like Web. So, I wou

Reviewer 2: I was talking about mobile, yeah. Thanks for the explanation!

```javascript
// For unhandled promise rejections
self.addEventListener('unhandledrejection', (error) => {
  console.warn('event unhandledrejection', error);
});

// For other exceptions
self.addEventListener('error', (error) => {
  console.warn('event error', error);
});
```

## Reporting errors

> It's possible to use 3rd party libaries for reporting errors like Bugsnag or Sentry. However, it might require some extra configs or setup depending on which libraries you are using because UI extensions are run inside WebWorker.

Reviewer 3: ```suggestion
It's possible to use 3rd party libaries like Bugsnag or Sentry for reporting errors. However, it might require extra configuration or setup, depending on which libraries you use, because UI extensions run inside [WebWorkers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API).
```

> It's possible to use 3rd party libaries like Bugsnag or Sentry for reporting errors. However, it might require extra configuration or setup, depending on which libraries you use, because UI extensions run inside [WebWorkers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API).

Reviewer 2: ```suggestion
It's possible to use 3rd party libaries like Bugsnag or Sentry for reporting errors. However, it might require extra configuration or setup, depending on which libraries you use, because UI extensions run inside [Web Workers](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API).
```

### Bugsnag

> Follow [Bugsnag documentation](https://docs.bugsnag.com/platforms/javascript/) to install it to your extension. If you use [CDN version](https://docs.bugsnag.com/platforms/javascript/cdn-guide/), you need to add polyfill code below as Bugsnag tries to access some variables that are not available in WebWorker.

Reviewer 3: ```suggestion
Follow [Bugsnag's documentation](https://docs.bugsnag.com/platforms/javascript/) to install it into your extension. If you use a [CDN version](https://docs.bugsnag.com/platforms/javascript/cdn-guide/), you need to add polyfill code below as Bugsnag tries to access some variables that are not available in WebWorkers.
```

```javascript
/**
 * CDN polyfill
 */
(() => {
  const document = {
    documentElement: {
      outerHTML: '',
      createElement: () => ({}),
      clientWidth: 0,
      clientHeight: 0,
    },
    addEventListener: () => {},
  };
  const history = {};
  const window = self;
  self.window = window;
  window.document = document;
  window.history = history;
})();

> ```

Reviewer 3: I know Chris mentioned it earlier but, yeah, this is a bug. If you get a chance, please file an issue for Bugsnag and link to it here (if possible) so that someone can see if it's fixed. 

This polyfill code seems fragile since a future version of Bugsnag could access different DOM properties.

Reviewer 2: Did we file an issue for this?

Reviewer 2: Also, do we think it's obvious enough where this code would go in an extension file? We don't show any context of what file/ where in a file we are adding this code, but not sure if we need to...

There is no need to add custom event listeners as Bugsnag does it for you out of the box.

### Sentry

> Follow [Sentry documentation](https://docs.sentry.io/platforms/javascript/) to install it to your extension. There is no polyfill required for using Sentry. However, Sentry does not add any event listeners out of the box. So, you need to add them manually.

Reviewer 3: ```suggestion
Follow [Sentry's documentation](https://docs.sentry.io/platforms/javascript/) to install it into your extension. A polyfill is not required for using Sentry. However, Sentry does not add any event listeners out of the box. So, you need to add them manually.
```

> Follow [Sentry's documentation](https://docs.sentry.io/platforms/javascript/) to install it into your extension. A polyfill is not required for using Sentry. However, Sentry does not add any event listeners out of the box. So, you need to add them manually.

Reviewer 2: Sentry also has a React-specific library we can reference: https://docs.sentry.io/platforms/javascript/guides/react/
